File: packages/go/pkg/types/rich_text.go
package types

// RichText models Telegram's RichText union: either a plain string for
// unformatted text, a single rich text entity (any of the RichText* structs in
// this file), or a []RichText of those two shapes.
//
// Docs spell it out as "either a String for plain text, an Array of RichText,
// or any of the following types", so no single Go type covers all three
// shapes; RichText is the escape hatch that keeps the wire format exact. Use
// RichTextEntity below to accept only the entity branch.
//
// Example:
//
//	var plain types.RichText = "Hello"
//	var entity types.RichText = types.RichTextBold{Type: "bold", Text: "world"}
//	var mixed types.RichText = []types.RichText{plain, entity}
//
// Telegram API: https://core.telegram.org/bots/api#richtext
type RichText any

// RichTextEntity is the interface implemented by every concrete RichText*
// variant in this file, so a handler can accept any entity regardless of
// format and still be type-checked, e.g.
//
//	var bold types.RichTextEntity = types.RichTextBold{Type: "bold", Text: "hi"}
//
// The concrete struct's Type field carries the wire discriminator documented
// on each variant.
//
// Telegram API: https://core.telegram.org/bots/api#richtext
type RichTextEntity interface {
	richTextEntity()
}

// RichTextAnchor represents an anchor.
//
// Telegram API: https://core.telegram.org/bots/api#richtextanchor
type RichTextAnchor struct {
	// Type of the rich text, always "anchor".
	Type string `json:"type"`
	// The name of the anchor.
	Name string `json:"name"`
}

func (RichTextAnchor) richTextEntity() {}

// RichTextAnchorLink represents a link to an anchor.
//
// Telegram API: https://core.telegram.org/bots/api#richtextanchorlink
type RichTextAnchorLink struct {
	// Type of the rich text, always "anchor_link".
	Type string `json:"type"`
	// The link text.
	Text RichText `json:"text"`
	// The name of the anchor. If the name is empty, then the link brings back to the top of the message.
	AnchorName string `json:"anchor_name"`
}

func (RichTextAnchorLink) richTextEntity() {}

// RichTextBankCardNumber represents a text with a bank card number.
//
// Telegram API: https://core.telegram.org/bots/api#richtextbankcardnumber
type RichTextBankCardNumber struct {
	// Type of the rich text, always "bank_card_number".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The bank card number.
	BankCardNumber string `json:"bank_card_number"`
}

func (RichTextBankCardNumber) richTextEntity() {}

// RichTextBold represents a bold text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextbold
type RichTextBold struct {
	// Type of the rich text, always "bold".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextBold) richTextEntity() {}

// RichTextBotCommand represents a bot command.
//
// Telegram API: https://core.telegram.org/bots/api#richtextbotcommand
type RichTextBotCommand struct {
	// Type of the rich text, always "bot_command".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The bot command.
	BotCommand string `json:"bot_command"`
}

func (RichTextBotCommand) richTextEntity() {}

// RichTextButton represents a button.
//
// Telegram API: https://core.telegram.org/bots/api#richtextbutton
type RichTextButton struct {
	// Type of the rich text, always "button".
	Type string `json:"type"`
	// The button.
	Button RichMessageButton `json:"button"`
}

func (RichTextButton) richTextEntity() {}

// RichTextCashtag represents a cashtag.
//
// Telegram API: https://core.telegram.org/bots/api#richtextcashtag
type RichTextCashtag struct {
	// Type of the rich text, always "cashtag".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The cashtag.
	Cashtag string `json:"cashtag"`
}

func (RichTextCashtag) richTextEntity() {}

// RichTextCode represents a monowidth text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextcode
type RichTextCode struct {
	// Type of the rich text, always "code".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextCode) richTextEntity() {}

// RichTextCustomEmoji represents a custom emoji.
//
// Telegram API: https://core.telegram.org/bots/api#richtextcustomemoji
type RichTextCustomEmoji struct {
	// Type of the rich text, always "custom_emoji".
	Type string `json:"type"`
	// Unique identifier of the custom emoji. Use getCustomEmojiStickers to get full information about the sticker.
	CustomEmojiID string `json:"custom_emoji_id"`
	// Alternative emoji for the custom emoji.
	AlternativeText string `json:"alternative_text"`
}

func (RichTextCustomEmoji) richTextEntity() {}

// RichTextDateTime represents formatted date and time.
//
// Telegram API: https://core.telegram.org/bots/api#richtextdatetime
type RichTextDateTime struct {
	// Type of the rich text, always "date_time".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The Unix time associated with the entity.
	UnixTime int64 `json:"unix_time"`
	// The string that defines the formatting of the date and time. See date-time entity formatting for more details.
	DateTimeFormat string `json:"date_time_format"`
}

func (RichTextDateTime) richTextEntity() {}

// RichTextEmailAddress represents a text with an email address.
//
// Telegram API: https://core.telegram.org/bots/api#richtextemailaddress
type RichTextEmailAddress struct {
	// Type of the rich text, always "email_address".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The email address.
	EmailAddress string `json:"email_address"`
}

func (RichTextEmailAddress) richTextEntity() {}

// RichTextHashtag represents a hashtag.
//
// Telegram API: https://core.telegram.org/bots/api#richtexthashtag
type RichTextHashtag struct {
	// Type of the rich text, always "hashtag".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The hashtag.
	Hashtag string `json:"hashtag"`
}

func (RichTextHashtag) richTextEntity() {}

// RichTextItalic represents an italicized text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextitalic
type RichTextItalic struct {
	// Type of the rich text, always "italic".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextItalic) richTextEntity() {}

// RichTextMarked represents a marked text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextmarked
type RichTextMarked struct {
	// Type of the rich text, always "marked".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextMarked) richTextEntity() {}

// RichTextMathematicalExpression represents a mathematical expression.
//
// Telegram API: https://core.telegram.org/bots/api#richtextmathematicalexpression
type RichTextMathematicalExpression struct {
	// Type of the rich text, always "mathematical_expression".
	Type string `json:"type"`
	// The expression in LaTeX format.
	Expression string `json:"expression"`
}

func (RichTextMathematicalExpression) richTextEntity() {}

// RichTextMention represents a mention by a username.
//
// Telegram API: https://core.telegram.org/bots/api#richtextmention
type RichTextMention struct {
	// Type of the rich text, always "mention".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The username.
	Username string `json:"username"`
}

func (RichTextMention) richTextEntity() {}

// RichTextPhoneNumber represents a text with a phone number.
//
// Telegram API: https://core.telegram.org/bots/api#richtextphonenumber
type RichTextPhoneNumber struct {
	// Type of the rich text, always "phone_number".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The phone number.
	PhoneNumber string `json:"phone_number"`
}

func (RichTextPhoneNumber) richTextEntity() {}

// RichTextReference represents a reference.
//
// Telegram API: https://core.telegram.org/bots/api#richtextreference
type RichTextReference struct {
	// Type of the rich text, always "reference".
	Type string `json:"type"`
	// Text of the reference.
	Text RichText `json:"text"`
	// The name of the reference.
	Name string `json:"name"`
}

func (RichTextReference) richTextEntity() {}

// RichTextReferenceLink represents a link to a reference.
//
// Telegram API: https://core.telegram.org/bots/api#richtextreferencelink
type RichTextReferenceLink struct {
	// Type of the rich text, always "reference_link".
	Type string `json:"type"`
	// The link text.
	Text RichText `json:"text"`
	// The name of the reference.
	ReferenceName string `json:"reference_name"`
}

func (RichTextReferenceLink) richTextEntity() {}

// RichTextSpoiler represents a text covered by a spoiler.
//
// Telegram API: https://core.telegram.org/bots/api#richtextspoiler
type RichTextSpoiler struct {
	// Type of the rich text, always "spoiler".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextSpoiler) richTextEntity() {}

// RichTextStrikethrough represents a strikethrough text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextstrikethrough
type RichTextStrikethrough struct {
	// Type of the rich text, always "strikethrough".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextStrikethrough) richTextEntity() {}

// RichTextSubscript represents a subscript text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextsubscript
type RichTextSubscript struct {
	// Type of the rich text, always "subscript".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextSubscript) richTextEntity() {}

// RichTextSuperscript represents a superscript text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextsuperscript
type RichTextSuperscript struct {
	// Type of the rich text, always "superscript".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextSuperscript) richTextEntity() {}

// RichTextTextMention represents a mention of a Telegram user by their identifier.
//
// Telegram API: https://core.telegram.org/bots/api#richtexttextmention
type RichTextTextMention struct {
	// Type of the rich text, always "text_mention".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// The mentioned user.
	User User `json:"user"`
}

func (RichTextTextMention) richTextEntity() {}

// RichTextUnderline represents an underlined text.
//
// Telegram API: https://core.telegram.org/bots/api#richtextunderline
type RichTextUnderline struct {
	// Type of the rich text, always "underline".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
}

func (RichTextUnderline) richTextEntity() {}

// RichTextUrl represents a text with a link.
//
// Telegram API: https://core.telegram.org/bots/api#richtexturl
type RichTextUrl struct {
	// Type of the rich text, always "url".
	Type string `json:"type"`
	// The text.
	Text RichText `json:"text"`
	// URL of the link.
	URL string `json:"url"`
}

func (RichTextUrl) richTextEntity() {}

// RichMessageButton represents a button in a RichMessage. Exactly one of the fields other than text and style must be used to specify the type of the button.
//
// Telegram API: https://core.telegram.org/bots/api#richmessagebutton
type RichMessageButton struct {
	// Text of the button. May contain only plain text, RichTextCustomEmoji and RichTextDateTime entities.
	Text RichText `json:"text"`
	// Style of the button. Must be one of "danger", "success", "primary", or "link" (the button is shown as a regular link without borders). Apps may use theme-specific colors for the button background and text based on the style. See the official docs for the full note.
	Style *string `json:"style,omitempty"`
	// HTTP or tg:// URL to be opened when the button is pressed. Links tg://user?id=<user_id> can be used to mention a user by their identifier without using a username, if this is allowed by their privacy settings.
	URL *string `json:"url,omitempty"`
	// Data to be sent in a callback query to the bot when the button is pressed, 1-64 bytes.
	CallbackData *string `json:"callback_data,omitempty"`
	// Description of the Web App that will be launched when the user presses the button. The Web App will be able to send an arbitrary message on behalf of the user using the method answerWebAppQuery. See the official docs for the full note.
	WebApp *WebAppInfo `json:"web_app,omitempty"`
	// An HTTPS URL used to automatically authorize the user. Can be used as a replacement for the Telegram Login Widget. Not supported for ephemeral messages. Telegram types this field as LoginUrl; pass a LoginUrl value. Go leaves this untyped because the LoginUrl model is not part of the rich-message types in this package.
	LoginURL any `json:"login_url,omitempty"`
	// If set, pressing the button will prompt the user to select one of their chats, open that chat and insert the bot's username and the specified inline query in the input field. See the official docs for the full note.
	SwitchInlineQuery *string `json:"switch_inline_query,omitempty"`
	// If set, pressing the button will insert the bot's username and the specified inline query in the current chat's input field. May be empty, in which case only the bot's username will be inserted. See the official docs for the full note.
	SwitchInlineQueryCurrentChat *string `json:"switch_inline_query_current_chat,omitempty"`
	// If set, pressing the button will prompt the user to select one of their chats of the specified type, open that chat and insert the bot's username and the specified inline query in the input field. See the official docs for the full note. Telegram types this field as SwitchInlineQueryChosenChat; pass a SwitchInlineQueryChosenChat value. Go leaves this untyped because the SwitchInlineQueryChosenChat model is not part of the rich-message types in this package.
	SwitchInlineQueryChosenChat any `json:"switch_inline_query_chosen_chat,omitempty"`
	// A button that copies the specified text to the clipboard. Telegram types this field as CopyTextButton; pass a CopyTextButton value. Go leaves this untyped because the CopyTextButton model is not part of the rich-message types in this package.
	CopyText any `json:"copy_text,omitempty"`
	// If set, then the button is disabled and does nothing. Telegram types this field as DisabledButton; pass a DisabledButton value. Go leaves this untyped because the DisabledButton model is not part of the rich-message types in this package.
	Disabled any `json:"disabled,omitempty"`
}

File: packages/python/src/telebot_py/types/rich_text.py
"""Rich formatted text nodes and rich message buttons (Bot API 10.3).

Field names, required-ness and ``type`` literals follow the official Telegram
Bot API docs for the ``RichText*`` classes and ``RichMessageButton``.

Every node class defaults its ``type`` field to the literal the docs give
("always 'bold'"), so callers never repeat it while :meth:`to_dict` always
emits it -- the same treatment ``remove_keyboard`` and ``force_reply`` get in
:mod:`telebot_py.types.keyboards`. Formatting nodes all carry one ``text``
field, "the text"; only fields whose meaning is not the attribute name are
listed in a class's ``Attributes:`` block.

``RichText`` models Telegram's three-way text shape: a bare ``str`` for plain
text, a sequence of nodes, or one node object. Optional fields use ``None``
for "not sent" (dropped by :meth:`to_dict`); an empty string or list is sent
as-is, so the ``UNSET`` sentinel is not needed on these constructors.

Remarks:
    ``RichMessageButton`` lives here with the text nodes because
    :class:`RichTextButton` embeds a button while a button's own ``text`` is a
    ``RichText``: the two are mutually recursive, and :meth:`from_dict`
    resolves annotations through the defining module's globals, so the union
    and both classes cannot be split across modules.
"""

from __future__ import annotations

import dataclasses
from collections.abc import Sequence

from telebot_py.types.base import TelegramObject
from telebot_py.types.keyboards import (
    CopyTextButton,
    LoginUrl,
    SwitchInlineQueryChosenChat,
    WebAppInfo,
)
from telebot_py.types.user import User


@dataclasses.dataclass(frozen=True, slots=True)
class RichMessageButton(TelegramObject):
    """A button in a rich message; exactly one field other than ``text`` and
    ``style`` chooses the button's type.

    Attributes:
        text: Text of the button; plain text, RichTextCustomEmoji or
            RichTextDateTime entities only.
        style: Style of the button; 'danger', 'success', 'primary' or 'link'.
        url: HTTP or tg:// URL opened when the button is pressed.
        callback_data: Data sent in a callback query when pressed, 1-64 bytes.
        web_app: Description of the Web App launched when pressed.
        login_url: An HTTPS URL used to automatically authorize the user.
        switch_inline_query: Prompts chat selection and inserts this query.
        switch_inline_query_current_chat: Inserts this query in the current chat.
        switch_inline_query_chosen_chat: Prompts selection of chats of these types.
        copy_text: A button that copies the given text to the clipboard.
        disabled: If set, the button is disabled and does nothing; kept raw
            because the Bot API ``DisabledButton`` has no fields.

    Telegram API: https://core.telegram.org/bots/api#richmessagebutton
    """

    text: RichText
    style: str | None = None
    url: str | None = None
    callback_data: str | None = None
    web_app: WebAppInfo | None = None
    login_url: LoginUrl | None = None
    switch_inline_query: str | None = None
    switch_inline_query_current_chat: str | None = None
    switch_inline_query_chosen_chat: SwitchInlineQueryChosenChat | None = None
    copy_text: CopyTextButton | None = None
    disabled: object | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextBold(TelegramObject):
    """A bold text; ``type`` is always ``'bold'``.

    Telegram API: https://core.telegram.org/bots/api#richtextbold
    """

    text: RichText
    type: str = "bold"
    _DISCRIMINATOR = ("type", "bold")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextItalic(TelegramObject):
    """An italicized text; ``type`` is always ``'italic'``.

    Telegram API: https://core.telegram.org/bots/api#richtextitalic
    """

    text: RichText
    type: str = "italic"
    _DISCRIMINATOR = ("type", "italic")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextUnderline(TelegramObject):
    """An underlined text; ``type`` is always ``'underline'``.

    Telegram API: https://core.telegram.org/bots/api#richtextunderline
    """

    text: RichText
    type: str = "underline"
    _DISCRIMINATOR = ("type", "underline")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextStrikethrough(TelegramObject):
    """A strikethrough text; ``type`` is always ``'strikethrough'``.

    Telegram API: https://core.telegram.org/bots/api#richtextstrikethrough
    """

    text: RichText
    type: str = "strikethrough"
    _DISCRIMINATOR = ("type", "strikethrough")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextSpoiler(TelegramObject):
    """A text covered by a spoiler; ``type`` is always ``'spoiler'``.

    Telegram API: https://core.telegram.org/bots/api#richtextspoiler
    """

    text: RichText
    type: str = "spoiler"
    _DISCRIMINATOR = ("type", "spoiler")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextSubscript(TelegramObject):
    """A subscript text; ``type`` is always ``'subscript'``.

    Telegram API: https://core.telegram.org/bots/api#richtextsubscript
    """

    text: RichText
    type: str = "subscript"
    _DISCRIMINATOR = ("type", "subscript")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextSuperscript(TelegramObject):
    """A superscript text; ``type`` is always ``'superscript'``.

    Telegram API: https://core.telegram.org/bots/api#richtextsuperscript
    """

    text: RichText
    type: str = "superscript"
    _DISCRIMINATOR = ("type", "superscript")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextMarked(TelegramObject):
    """A marked text; ``type`` is always ``'marked'``.

    Telegram API: https://core.telegram.org/bots/api#richtextmarked
    """

    text: RichText
    type: str = "marked"
    _DISCRIMINATOR = ("type", "marked")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextCode(TelegramObject):
    """A monowidth text; ``type`` is always ``'code'``.

    Telegram API: https://core.telegram.org/bots/api#richtextcode
    """

    text: RichText
    type: str = "code"
    _DISCRIMINATOR = ("type", "code")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextDateTime(TelegramObject):
    """Formatted date and time; ``type`` is always ``'date_time'``.

    Attributes:
        unix_time: The Unix time associated with the entity.
        date_time_format: The string defining how the date and time are shown.

    Telegram API: https://core.telegram.org/bots/api#richtextdatetime
    """

    text: RichText
    unix_time: int
    date_time_format: str
    type: str = "date_time"
    _DISCRIMINATOR = ("type", "date_time")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextTextMention(TelegramObject):
    """A mention of a Telegram user by their identifier; ``type`` is always
    ``'text_mention'``.

    Attributes:
        user: The mentioned user.

    Telegram API: https://core.telegram.org/bots/api#richtexttextmention
    """

    text: RichText
    user: User
    type: str = "text_mention"
    _DISCRIMINATOR = ("type", "text_mention")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextCustomEmoji(TelegramObject):
    """A custom emoji; ``type`` is always ``'custom_emoji'``.

    Attributes:
        custom_emoji_id: Unique identifier of the custom emoji.
        alternative_text: Alternative emoji for the custom emoji.

    Telegram API: https://core.telegram.org/bots/api#richtextcustomemoji
    """

    custom_emoji_id: str
    alternative_text: str
    type: str = "custom_emoji"
    _DISCRIMINATOR = ("type", "custom_emoji")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextMathematicalExpression(TelegramObject):
    """A mathematical expression; ``type`` is always
    ``'mathematical_expression'``.

    Attributes:
        expression: The expression in LaTeX format.

    Telegram API: https://core.telegram.org/bots/api#richtextmathematicalexpression
    """

    expression: str
    type: str = "mathematical_expression"
    _DISCRIMINATOR = ("type", "mathematical_expression")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextUrl(TelegramObject):
    """A text with a URL; ``type`` is always ``'url'``.

    Attributes:
        url: The URL.

    Telegram API: https://core.telegram.org/bots/api#richtexturl
    """

    text: RichText
    url: str
    type: str = "url"
    _DISCRIMINATOR = ("type", "url")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextEmailAddress(TelegramObject):
    """A text with an email address; ``type`` is always ``'email_address'``.

    Attributes:
        email_address: The email address.

    Telegram API: https://core.telegram.org/bots/api#richtextemailaddress
    """

    text: RichText
    email_address: str
    type: str = "email_address"
    _DISCRIMINATOR = ("type", "email_address")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextPhoneNumber(TelegramObject):
    """A text with a phone number; ``type`` is always ``'phone_number'``.

    Attributes:
        phone_number: The phone number.

    Telegram API: https://core.telegram.org/bots/api#richtextphonenumber
    """

    text: RichText
    phone_number: str
    type: str = "phone_number"
    _DISCRIMINATOR = ("type", "phone_number")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextBankCardNumber(TelegramObject):
    """A text with a bank card number; ``type`` is always
    ``'bank_card_number'``.

    Attributes:
        bank_card_number: The bank card number.

    Telegram API: https://core.telegram.org/bots/api#richtextbankcardnumber
    """

    text: RichText
    bank_card_number: str
    type: str = "bank_card_number"
    _DISCRIMINATOR = ("type", "bank_card_number")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextMention(TelegramObject):
    """A mention by a username; ``type`` is always ``'mention'``.

    Attributes:
        username: The username.

    Telegram API: https://core.telegram.org/bots/api#richtextmention
    """

    text: RichText
    username: str
    type: str = "mention"
    _DISCRIMINATOR = ("type", "mention")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextHashtag(TelegramObject):
    """A hashtag; ``type`` is always ``'hashtag'``.

    Attributes:
        hashtag: The hashtag.

    Telegram API: https://core.telegram.org/bots/api#richtexthashtag
    """

    text: RichText
    hashtag: str
    type: str = "hashtag"
    _DISCRIMINATOR = ("type", "hashtag")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextCashtag(TelegramObject):
    """A cashtag; ``type`` is always ``'cashtag'``.

    Attributes:
        cashtag: The cashtag.

    Telegram API: https://core.telegram.org/bots/api#richtextcashtag
    """

    text: RichText
    cashtag: str
    type: str = "cashtag"
    _DISCRIMINATOR = ("type", "cashtag")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextBotCommand(TelegramObject):
    """A bot command; ``type`` is always ``'bot_command'``.

    Attributes:
        bot_command: The bot command.

    Telegram API: https://core.telegram.org/bots/api#richtextbotcommand
    """

    text: RichText
    bot_command: str
    type: str = "bot_command"
    _DISCRIMINATOR = ("type", "bot_command")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextButton(TelegramObject):
    """A button; ``type`` is always ``'button'``.

    Attributes:
        button: The button.

    Telegram API: https://core.telegram.org/bots/api#richtextbutton
    """

    button: RichMessageButton
    type: str = "button"
    _DISCRIMINATOR = ("type", "button")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextAnchor(TelegramObject):
    """An anchor; ``type`` is always ``'anchor'``.

    Attributes:
        name: The name of the anchor.

    Telegram API: https://core.telegram.org/bots/api#richtextanchor
    """

    name: str
    type: str = "anchor"
    _DISCRIMINATOR = ("type", "anchor")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextAnchorLink(TelegramObject):
    """A link to an anchor; ``type`` is always ``'anchor_link'``.

    Attributes:
        anchor_name: Name of the anchor; empty brings the reader to the top.

    Telegram API: https://core.telegram.org/bots/api#richtextanchorlink
    """

    text: RichText
    anchor_name: str
    type: str = "anchor_link"
    _DISCRIMINATOR = ("type", "anchor_link")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextReference(TelegramObject):
    """A reference; ``type`` is always ``'reference'``.

    Attributes:
        name: The name of the reference.

    Telegram API: https://core.telegram.org/bots/api#richtextreference
    """

    text: RichText
    name: str
    type: str = "reference"
    _DISCRIMINATOR = ("type", "reference")


@dataclasses.dataclass(frozen=True, slots=True)
class RichTextReferenceLink(TelegramObject):
    """A link to a reference; ``type`` is always ``'reference_link'``.

    Attributes:
        reference_name: The name of the reference.

    Telegram API: https://core.telegram.org/bots/api#richtextreferencelink
    """

    text: RichText
    reference_name: str
    type: str = "reference_link"
    _DISCRIMINATOR = ("type", "reference_link")


#: A rich text entity node, excluding plain text.
RichTextNode = (
    RichTextBold
    | RichTextItalic
    | RichTextUnderline
    | RichTextStrikethrough
    | RichTextSpoiler
    | RichTextDateTime
    | RichTextTextMention
    | RichTextSubscript
    | RichTextSuperscript
    | RichTextMarked
    | RichTextCode
    | RichTextCustomEmoji
    | RichTextMathematicalExpression
    | RichTextUrl
    | RichTextEmailAddress
    | RichTextPhoneNumber
    | RichTextBankCardNumber
    | RichTextMention
    | RichTextHashtag
    | RichTextCashtag
    | RichTextBotCommand
    | RichTextButton
    | RichTextAnchor
    | RichTextAnchorLink
    | RichTextReference
    | RichTextReferenceLink
)

#: A rich formatted text: a bare ``str`` for plain text, a sequence of nodes,
#: or a single entity node.
RichText = str | RichTextNode | Sequence[str | RichTextNode]
